scenario_player/utils/reclaim.py

```diff
@@ -0,0 +1,425 @@
+    for node in reclamation_candidates:
+        balance = web3.eth.getBalance(node.address)
+        if balance > reclaim_tx_cost:
+            drain_amount = balance - reclaim_tx_cost
+            log.info("Reclaiming", from_address=node.address, amount=drain_amount.__format__(",d"))
+            reclaim_amount += drain_amount
+            client = JSONRPCClient(web3, node.privkey)
+            assert account.address
+            txs.append(
+                client.transact(
+                    EthTransfer(
+                        to_address=account.address, value=drain_amount, gas_price=gas_price
+                    )
+                )
+            )
+        (node.node_dir / "reclaimed").touch()
+
+    wait_for_txs(web3, txs, 1000)
+    log.info("Reclaimed", reclaim_amount=reclaim_amount.__format__(",d"))
+
+
+def _get_channels(
+    candidate: ReclamationCandidate,
+    contract_manager: ContractManager,
+    web3: Web3,
+    token_network_address: TokenNetworkAddress,
+    deploy: DeployedContracts,
+) -> List[dict]:
+    """ Read ChannelOpened events to get the channel_ids and participants """
+    new_channel_abi = contract_manager.get_event_abi(CONTRACT_TOKEN_NETWORK, ChannelEvent.OPENED)
+    topics = [
+        encode_hex(event_abi_to_log_topic(new_channel_abi)),  # type: ignore
+        None,
+        encode_hex(bytes([0] * 12) + to_canonical_address(candidate.address)),
+    ]
+    events = query_blockchain_events(
+        web3=web3,
+        contract_manager=contract_manager,
+        contract_address=Address(token_network_address),
+        contract_name=CONTRACT_TOKEN_NETWORK,
+        topics=topics,
+        from_block=BlockNumber(
+            deploy["contracts"][CONTRACT_TOKEN_NETWORK_REGISTRY]["block_number"]
+        ),
+        to_block=web3.eth.blockNumber,
+    )
+    return [ev["args"] for ev in events]
+
+
+def _get_token_network_address(
+    token_address: TokenAddress, web3: Web3, privkey: PrivateKey, deploy: DeployedContracts
+) -> TokenNetworkAddress:
+
+    client = JSONRPCClient(web3, privkey)
+    proxy_manager = get_proxy_manager(client, deploy)
+    token_network_registry_address = TokenNetworkRegistryAddress(
+        to_canonical_address(deploy["contracts"][CONTRACT_TOKEN_NETWORK_REGISTRY]["address"])
+    )
+    token_network_registry = proxy_manager.token_network_registry(
+        token_network_registry_address, "latest"
+    )
+    token_network_address = token_network_registry.get_token_network(token_address, "latest")
+    assert token_network_address
+    return token_network_address
+
+
+def _withdraw_all_from_channel(
+    candidate: ReclamationCandidate,
+    reclamation_candidates: List[ReclamationCandidate],
+    channel: dict,
+    token_network: TokenNetwork,
+):
+    # Check if channel still has deposits
+    details = token_network.detail_participants(
+        participant1=to_canonical_address(candidate.address),
+        participant2=to_canonical_address(channel["participant2"]),
```

Comment:
There is no guarantee that `participant2` is not the *candidate* address. First one has to make sure the partner address is used instead, something like:

```python
if channel['participant1'] == candidate.address:
  partner_address = channel['participant1']
elif channel['participant2'] == candidate.address:
  partner_address = channel['participant1']
else:
  raise RuntimeError('not a participant of the given channel')
```

the here you can use `partner_address` instead of `channel["participant2"]`.

Reply:
The same applies to all `channel["participant2"]` below.

Reply:
The ChannelOpened events are filtered by `participant1 == candidate`. Since you can't open channels with `participant1 == participant2` ti should be guaranteed, right?

Reply:
Humm, that is correct, but I had no way of telling this was the intention from the code, maybe a comment to point this out next time?

With that said, the code still had the problem of not handling withdraws for `participant2`, which I suppose was not intentional? And then this function would not be allowed to have that assumption.

Reply:
All tokens are withdrawn by participant1, so there is no need to do a withdraw for participant2.  I tried to explain that in the docstring
>      All tokens will be withdrawn to participant1, ignoring all balance proofs. 
but apparently was not explicit enough.